torchdata/dataloader2/dataloader2.py

```diff
@@ -43,6 +43,38 @@ class ConcurrencySpec:
     persistent_workers: bool = False
 
 
+class _DataLoader2Iterator(Iterator):
+    def __init__(self, dataloader: "DataLoader2", iterator_id: int):
+        self.dataloader = dataloader
+        self.iterator_id = iterator_id
+
+    def __next__(self) -> T_co:
+        if self.iterator_id == self.dataloader.valid_iterator_id:
+            self.dataloader._reset_iter = True
```

Comment:
Why doe we need to `reset_iter` every single `next`? IIUC, you should implement a `__iter__` function with such logic and return `self`.

Reply:
`self.dataloader._reset_iter` determines whether reset needs to happen when `iter(DataLoader2)` is called, so this is not "reset_iter" everything single `next`. 

Ideally, `self.dataloader._reset_iter = True` should only be set once the first time `next` is called. We can choose to set it once when `iter` is called (instead of the first time `next` is called). but that may lead to some unnecessary reset of `DataLoader2`.

Reply:
I see. Then, the question is do we need `_reset_iter`?
When I created this `_reset_iter` flag, it's used for mini-epoch. Only when `PauseIteration` is raised, we don't want to reset datapipe iterator but re-use the datapipe iterator since it's not depleted yet. Could we still achieve this behavior?

Reply:
Discussed offline

Reply:
Same question as @ejguan, are we still allowing resuing iterator when receiving a `PauseIteration`? We can use internal unit tests to verify that. 
For context, our reading service may need to support both mini-epoch and real multiple epochs training soon. But that should be a reading-service update to raise Pause/StopIteration based on some flag.

Reply:
Yes, I expect the iterator can be used as long as a new one isn't created. We should verify with internal test.